src/Components/Blazor/WebAssembly.Authentication/src/AuthenticationManager.cs

```diff
@@ -0,0 +1,446 @@
+                throw new InvalidOperationException("Invalid return url. The return url needs to have the same origin as the current page.");
+            }
+
+            return state?.ReturnUrl ?? fromQuery ?? defaultReturnUrl ?? Navigation.BaseUri;
+        }
+
+        private string GetParameter(string key)
+        {
+            var queryString = new Uri(Navigation.Uri).Query;
+
+            if (string.IsNullOrEmpty(queryString) || queryString == "?")
+            {
+                return null;
+            }
+
+            var scanIndex = 0;
+            if (queryString[0] == '?')
+            {
+                scanIndex = 1;
+            }
+
+            var textLength = queryString.Length;
+            var equalIndex = queryString.IndexOf('=');
+            if (equalIndex == -1)
+            {
+                equalIndex = textLength;
+            }
+
+            while (scanIndex < textLength)
+            {
+                var ampersandIndex = queryString.IndexOf('&', scanIndex);
+                if (ampersandIndex == -1)
+                {
+                    ampersandIndex = textLength;
+                }
+
+                if (equalIndex < ampersandIndex)
+                {
+                    while (scanIndex != equalIndex && char.IsWhiteSpace(queryString[scanIndex]))
+                    {
+                        ++scanIndex;
+                    }
+                    var name = queryString[scanIndex..equalIndex];
+                    var value = queryString.Substring(equalIndex + 1, ampersandIndex - equalIndex - 1);
+                    var processedName = Uri.UnescapeDataString(name.Replace('+', ' '));
+                    if (string.Equals(processedName, key, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return Uri.UnescapeDataString(value.Replace('+', ' '));
+                    }
+
+                    equalIndex = queryString.IndexOf('=', ampersandIndex);
+                    if (equalIndex == -1)
+                    {
+                        equalIndex = textLength;
+                    }
+                }
+                else
+                {
+                    if (ampersandIndex > scanIndex)
+                    {
+                        var value = queryString[scanIndex..ampersandIndex];
+                        if (string.Equals(value, key, StringComparison.OrdinalIgnoreCase))
+                        {
+                            return string.Empty;
+                        }
+                    }
+                }
+
+                scanIndex = ampersandIndex + 1;
+            }
+
+            return null;
+        }
+
+        private async Task NavigateToReturnUrl(string returnUrl) => await JS.InvokeVoidAsync("location.replace", returnUrl);
+
+        private ValueTask RedirectToRegister()
+        {
+            var loginUrl = Navigation.ToAbsoluteUri(ApplicationPaths.LoginPath).PathAndQuery;
```

Comment:
I'm not sure it's necessary to go through this process of converting to an absolute URL and taking the "path and query" part. Won't it work just the same if we keep the base-relative URL and use that?

If the reason is so that we can check the origin matches later, what do you think about the alternative of *requiring* it to be a base-relative URL so there isn't even any way of describing an external redirection? I don't mind very strongly - just asking to clarify.

Reply:
The auth server might have a different base path so the loginUrl needs to be host relative. In general I prefer to use absolute urls to avoid confusion where possible and for safety against open redirects.